### ISSUE TITLE
tf dataset sample_weight is not passed to custom metrics (or any metrics).

### ISSUE THREAD

**Issue by kechan** [tensorflow/tensorflow#44068]
**System information**
- v2.3.0-0-gb36436b087 2.3.0
- code is tested on google colab

**Describe the current behavior**
sample_weight as defined in tf.data.Dataset (not passed explicitly in model.fit(...)) is not being called with 
any metrics' update_state(self, y_true, y_pred, sample_weight). 

**Describe the expected behavior**
sample_weight should be passed to update_state of metrics.

**Standalone code to reproduce the issue**
Colab Notebook
https://colab.research.google.com/drive/1eSVeRuUv0q_KbSHQ2C4AFSIep_HIt1mn?usp=sharing

```
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

x = np.random.randn(10, 1).astype('float32')
y = np.random.randn(10, 1).astype('float32')
w = np.random.randint(0, 2, (10, 1)).astype('float32')   # sample weight

ds = tf.data.Dataset.from_tensor_slices((x, y, w)).batch(2)

model = Sequential()   # simple linear regression
model.add(Dense(1))

# Custom metrics, we will just sum up the sample_weight

class TestSampleWeight(tf.keras.metrics.Metric):
  def __init__(self, name="test_sample_weight", **kwargs):
    super(TestSampleWeight, self).__init__(name=name, **kwargs)
    self.sample_weight_sum = self.add_weight(name="sws", initializer="zeros")

  def update_state(self, y_true, y_pred, sample_weight=None):
    y_pred = tf.reshape(tf.argmax(y_pred, axis=1), shape=(-1, 1))
    values = tf.cast(y_true, "int32") == tf.cast(y_pred, "int32")
    values = tf.cast(values, "float32")
    
    # self.sample_weight_sum.assign_add(1)    # Uncomment to simple sanity test for metrics, comment out next 2 lines.

    sample_weight = tf.cast(sample_weight, "float32")
    self.sample_weight_sum.assign_add(sample_weight)

  def result(self):
    return self.sample_weight_sum

  def reset_states(self):
    self.sample_weight_sum.assign(0.0)

model.compile(optimizer='rmsprop', loss='mse', metrics=TestSampleWeight())

model.fit(ds)  # hit ValueError: None values not supported. sample_weight is None?? 
```

**Other info / logs**

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/f9624842a1aca1df03a7003f419381e2/44068.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/f9b3d4319ed0f5888d65849aadced529/44068-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by kechan** [comment]
@amahendrakar Did you attach any gist yet? I actually am also able to reproduce this without using tf.data.Dataset, just using plain good old np.array passed into .fit() as x, y, and sample_weight. I am somewhat surprised this isn’t raised, unless it is a recent regression.

**Comment by kechan** [comment]
I think I spot the problem area from the stack trace (triggered by me intentionally operated on a None thats supposed to be the sample_weight).

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/compile_utils.py
     407         if metric_obj is None:
     408           continue
--> 409         metric_obj.update_state(y_t, y_p, sample_weight=mask)

If you look at the full code, there's another var named 'sw', which i think should be the sample_weight. I am totally unfamiliar with this code but I am guessing you want to pass 'sw' as sample_weight and not the 'mask'?? I debugged that 'mask' here is in fact a None and that obviously explain why sample_weight called in update_state is also None.

 I went ahead and replaced 'mask' with 'sw' on line 409 on my local. After that, the sample_weight seemed to be passed correctly onto update_state(...).
